### DISCUSSION TITLE
guides/kfold-cross-validation/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#13044] [Docs]
# guides/kfold-cross-validation/

An in-depth guide demonstrating the implementation of K-Fold Cross Validation with the Ultralytics ecosystem for object detection datasets, leveraging Python, YOLO, and sklearn.

https://docs.ultralytics.com/guides/kfold-cross-validation/

**Comment by fazrigading** [comment]
Is there any way to implement Stratified K-Fold Cross Validation with YOLOv8?

**Reply by pderrenger** [reply]
Yes, you can implement Stratified K-Fold Cross Validation with YOLOv8! Stratified K-Fold ensures that each fold has a similar distribution of class labels, which can be particularly useful for imbalanced datasets.

To achieve this, you can use `StratifiedKFold` from `sklearn.model_selection` instead of `KFold`. Here's a brief example of how to modify the K-Fold implementation:

```python
from sklearn.model_selection import StratifiedKFold

# Assuming labels_df is your DataFrame with class counts
ksplit = 5
skf = StratifiedKFold(n_splits=ksplit, shuffle=True, random_state=20)

# Generate stratified splits
kfolds = list(skf.split(labels_df, labels_df.idxmax(axis=1)))

# The rest of the process remains the same
```

For a detailed guide on K-Fold Cross Validation, please refer to the [Ultralytics documentation](https://docs.ultralytics.com/guides/kfold-cross-validation/). If you have any more questions, feel free to ask! 😊

**Comment by alexander-sony** [comment]
Good guide, consider adding it in a colab notebook.

In section "Train YOLO using K-Fold Data Splits", model should be re-initialized foe every k, i.e. it should be inside the for-loop: for k in range(ksplit).

**Reply by glenn-jocher** [reply]
@alexander-sony thank you for your feedback and suggestion! 😊

You're absolutely right about re-initializing the model within the loop for each fold to ensure a fresh start for every training iteration. This can help avoid any potential carryover effects from previous folds. Here's the corrected code snippet for the training section:

```python
from ultralytics import YOLO

weights_path = "path/to/weights.pt"
results = {}

# Define your additional arguments here
batch = 16
project = "kfold_demo"
epochs = 100

for k in range(ksplit):
    model = YOLO(weights_path, task="detect")  # Re-initialize the model inside the loop
    dataset_yaml = ds_yamls[k]
    model.train(data=dataset_yaml, epochs=epochs, batch=batch, project=project)  # include any train arguments
    results[k] = model.metrics  # save output metrics for further analysis
```

Regarding your suggestion to add this guide to a Colab notebook, that's a fantastic idea! It would make the process more interactive and accessible for users. We'll consider creating a Colab version of this guide in the future.

For more details on K-Fold Cross Validation, you can refer to our [comprehensive guide](https://docs.ultralytics.com/guides/kfold-cross-validation/).

Thanks again for your valuable input! If you have any more suggestions or questions, feel free to share. Happy coding! 🚀

**Reply by Likhith-Sugganahalli** [reply]
@alexander-sony

woudnt that be a new model trained on each split, giving us n different models trained on 1/nth of the dataset?

**Comment by kh-diab** [comment]
is it usefult to use this k fold approach when I just have one class only 
but my issue is the class is detected as background or vice versa as per the attached Confusion matrix

**Reply by glenn-jocher** [reply]
@kh-diab using K-Fold Cross Validation can still be beneficial even with a single class. It helps ensure that your model's performance is robust and not overly dependent on a particular subset of the data. This can be particularly useful if you're experiencing issues with class detection accuracy, as it allows you to validate the model's performance across multiple splits and identify any inconsistencies. If your class is being detected as background or vice versa, K-Fold Cross Validation can help diagnose whether this issue is consistent across different data splits or specific to certain subsets. For more details, you can refer to our guide on K-Fold Cross Validation: https://docs.ultralytics.com/guides/kfold-cross-validation/.

**Comment by semyon-v-dot** [comment]
After the loop:
```
for ext in supported_extensions:
    images.extend(sorted((dataset_path / "images").rglob(f"*{ext}")))
```

Should be this line of code:
```images = sorted(images)```

For `images` matching with `labels` list:
```labels = sorted(dataset_path.rglob("*labels/*.txt"))```

Otherwise the script creates background images because some images missing their labels.

**Reply by glenn-jocher** [reply]
@semyon-v-dot thank you for your suggestion. Sorting the `images` list after extending it with supported extensions is a good practice to ensure alignment with the `labels` list. This helps prevent issues with missing labels.

**Comment by Nehasq** [comment]
Hi, I am applying K-fold validation using the leave-one-out technique. My dataset consists of a single class and includes 21 images in total. For each split, I use 20 images for training and 1 image for testing. I am implementing the leave-one-out technique with the code above, setting ksplit=21.

I am not using a separate validation dataset, only training and testing data. Here is the code I am using:

results = {}

# Define additional arguments here
batch = 16
project = "kfold_demo"
epochs = 100

for k in range(ksplit):
    dataset_yaml = ds_yamls[k]
    model = YOLO(weights_path, task="detect")
    model.train(data=dataset_yaml, epochs=epochs, batch=batch, project=project)  # Include any train arguments
    results[k] = model.metrics  # Save output metrics for further analysis




In my data.yaml file, I have provided the validation path as the test set.
When it trains and validates on the "1 test data," does this mean the model is being tested on data it has never seen during training? This is my first point of confusion.

my data.yaml file for each split looks like this:
names:
- pothole
path: ../ 2024-12-10_21-Fold_Cross-val/split_1
train: train
val: test


My second question is about calculating the final performance. When I ran the above code, it calculated scores for each individual split. Do I need to calculate the final performance metrics manually, or is there any functionality in YOLOv8 to compute the aggregated performance automatically?

**Reply by glenn-jocher** [reply]
@Nehasq to address your first question, yes, the model is being tested on unseen data during each split, as the test set (1 image) is excluded from training and only used for validation. Regarding your second question, YOLO does not automatically aggregate metrics across splits. You need to compute the final performance metrics manually by averaging the results from all splits. For reference, you can review the [K-Fold Cross Validation guide](https://docs.ultralytics.com/guides/kfold-cross-validation/).

**Reply by Nehasq** [reply]
Thanks for your reply.
I understand your point. So, are you saying that, for example, in my case, where I have 21 scores (precision, recall, etc.), I should simply add all the precision values together and divide the sum by 21 to get the final precision of the model? Is that correct?

Also, in y case of 20 images( training) and 1 images for test, what batch size would you suggest?

**Reply by glenn-jocher** [reply]
Yes, to calculate the final precision (or any metric), you average the values across all splits, i.e., sum all precision scores and divide by the number of splits (21 in your case). As for the batch size, given your small dataset, you could keep it low (e.g., 1 or 2) to ensure efficient utilization of the data while avoiding overfitting.

**Reply by Nehasq** [reply]
Thanks, that really helps! With a batch size of 1, I am keeping the epochs at 10. Should it be fewer? Also, to check for overfitting, I need to review the result graphs for each split. However, I can't definitively determine if it is overfitting since each split is separate, right?

**Reply by glenn-jocher** [reply]
Keeping epochs at 10 with a batch size of 1 is reasonable given your small dataset size. To check for overfitting, reviewing result graphs from each split can help, but overfitting may be harder to detect since each split involves different training and test data. Consider comparing trends across splits for consistent insights. For further details, you can reference the [K-Fold Cross Validation guide](https://docs.ultralytics.com/guides/kfold-cross-validation/).

**Comment by jingieboy** [comment]
does this work with yolov8-cls? if so how would the folder structure look like? how are the data splits

**Reply by glenn-jocher** [reply]
Yes—K‑Fold works the same for classification; use the standard ImageNet-style layout per fold, and train each fold separately (YOLO infers classes from folder names). For YOLO11 (recommended) you can do:

```text
dataset/
  split_1/
    train/
      class_a/ ...
      class_b/ ...
    val/
      class_a/ ...
      class_b/ ...
  split_2/
    train/...
    val/...
  ...
```

```python
from ultralytics import YOLO

weights = "yolo11n-cls.pt"  # swap to 'yolov8n-cls.pt' if you prefer YOLOv8
splits = ["dataset/split_1", "dataset/split_2", "dataset/split_3"]

for d in splits:
    model = YOLO(weights)
    model.train(data=d, epochs=20, batch=16)  # d must contain 'train/' and 'val/' subfolders
```

For building folds, use `StratifiedKFold` on image-level labels derived from parent folder names to keep class balance, or use the random splitter shown in the [autosplit reference](https://docs.ultralytics.com/reference/data/split/) if stratification isn’t required. Classification usage details are in the [classification task docs](https://docs.ultralytics.com/tasks/classify/).

**Reply by jingieboy** [reply]
thank you for this! I have also notice that everytime I run model.val(), the accuracies and confusion matrix values are slightly different this time. Is this normal or during evaluation, there is a small randomness when calling the function.

**Reply by glenn-jocher** [reply]
@jingieboy minor variations can occur from non-deterministic CUDA ops or multi-threading; validation itself doesn’t add randomness by default—set seeds and force deterministic ops to lock results:
```python
import random, numpy as np, torch

seed = 0
random.seed(seed); np.random.seed(seed); torch.manual_seed(seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

model.val()  # run evaluation
```
If differences are still large, double-check your eval args in the [validation mode docs](https://docs.ultralytics.com/modes/val/).
